python/pyspark/sql/connect/session.py

```diff
@@ -231,13 +231,21 @@ def createDataFrame(
                         _schema.fields[i].name = name
                         _schema.names[i] = name
 
-            pdf = pd.DataFrame(_data)
-
             if _cols is None:
-                _cols = ["_%s" % i for i in range(1, pdf.shape[1] + 1)]
+                if _schema is None:
+                    _cols = ["_%s" % i for i in range(1, len(_data[0]) + 1)]
+                else:
+                    _cols = _schema.names
+
+            if isinstance(_data[0], Row):
+                table = pa.Table.from_pylist([row.asDict() for row in _data])
```

Comment:
Should probably call `row.asDict(recursive=True)`

Reply:
will update